File: apps/android/app/src/main/java/net/develivarr/auralis/features/search/SearchSuggestions.kt
package net.develivarr.auralis.features.search

/** Where selecting a [SearchSuggestionUi] navigates — the exact same target its kind's row in
 * the full results list below already navigates to (`docs/design/screens/SEARCH.md` §6.2's
 * "Selection" rule: "no new routes"). [TrackAlbum] is always constructed with a non-null
 * `albumId` — [deriveSearchSuggestions] excludes a track with none entirely, the same "nowhere
 * to go" rule §5 already applies to the full track list. */
sealed interface SearchSuggestionTarget {
    data class Book(val id: String) : SearchSuggestionTarget

    data class Podcast(val id: String) : SearchSuggestionTarget

    data class Artist(val id: String) : SearchSuggestionTarget

    data class Album(val id: String) : SearchSuggestionTarget

    data class TrackAlbum(val albumId: String) : SearchSuggestionTarget
}

/**
 * One typeahead suggestion (`docs/design/screens/SEARCH.md` §6.2). [title] is the plain,
 * undecorated title — what selecting the suggestion writes back into the search field, per
 * §6.2's "Selection" rule ("the suggestion's plain title (not the decorated '· Kind' label)").
 * [label] is the full `"{title} · {Kind}"` display string (§6.2), built once here rather than at
 * each render site so the dropdown item and its `contentDescription` (§11) always agree by
 * construction.
 */
data class SearchSuggestionUi(
    val id: String,
    val title: String,
    val label: String,
    val target: SearchSuggestionTarget,
)

/** §3's Suggestion cap — 8 total, across all kinds combined, a product decision made once in
 * the spec with no Sonora source. */
internal const val SEARCH_SUGGESTION_CAP = 8

/**
 * Derives up to [SEARCH_SUGGESTION_CAP] typeahead suggestions (§6.2) from the same
 * `books`/`podcasts`/`artists`/`albums`/`tracks` arrays [UnifiedSearchResultsUiState.Results]
 * already carries — no new fetch, no new debounce, updating on exactly the results list's own
 * cadence, matching §6.2's "Source" rule.
 *
 * Kind order: Books, Podcasts, Artists, Albums, Tracks — the same fixed order
 * [UnifiedSearchScreen]'s own results list already renders in, per §6.2's "Ordering and cap"
 * rule, minus the two kinds this function deliberately excludes:
 *
 * **Series and Authors are excluded entirely on Android** — the one deliberately unequal point
 * in §6.2. Neither kind has a detail route on this platform yet (confirmed absent from
 * `AuralisNavHost.kt`'s `Routes` object), and per §5/§6.2 an inert suggestion — one that cannot
 * be navigated to — is a worse version of the inert-row problem the full results list already
 * solves by rendering those two kinds as plain, non-interactive text. This is *idiom*, not
 * drift: it is forced by a route that genuinely does not exist yet, not by this wave declining
 * to build one. Excluding these two kinds entirely (rather than counting empty slots for them)
 * is why a platform missing a kind's route still gets up to [SEARCH_SUGGESTION_CAP] suggestions
 * from the kinds it does have, per §6.2's own wording.
 *
 * A track with no [net.develivarr.auralis.features.music.MusicSearchTrackUi.albumId] is excluded
 * too — same "nowhere to go" rule §5 already applies to the full track list, and the same reason
 * [SearchSuggestionTarget.TrackAlbum] never carries a null `albumId`.
 *
 * Requestable-books/requestable-music candidates are never considered — this function only ever
 * reads [UnifiedSearchResultsUiState.Results], never [RequestableBooksUiState]/
 * [RequestableMusicUiState], so §6.2's "settles on its own, slower, debounced cadence" exclusion
 * holds by construction rather than needing an explicit check.
 */
internal fun deriveSearchSuggestions(state: UnifiedSearchResultsUiState.Results): List<SearchSuggestionUi> {
    val suggestions = mutableListOf<SearchSuggestionUi>()

    for (book in state.books) {
        if (suggestions.size >= SEARCH_SUGGESTION_CAP) return suggestions
        suggestions +=
            SearchSuggestionUi(
                id = "book:${book.id}",
                title = book.title,
                label = "${book.title} · Book",
                target = SearchSuggestionTarget.Book(book.id),
            )
    }
    for (podcast in state.podcasts) {
        if (suggestions.size >= SEARCH_SUGGESTION_CAP) return suggestions
        suggestions +=
            SearchSuggestionUi(
                id = "podcast:${podcast.id}",
                title = podcast.title,
                label = "${podcast.title} · Podcast",
                target = SearchSuggestionTarget.Podcast(podcast.id),
            )
    }
    for (artist in state.artists) {
        if (suggestions.size >= SEARCH_SUGGESTION_CAP) return suggestions
        suggestions +=
            SearchSuggestionUi(
                id = "artist:${artist.id}",
                title = artist.name,
                label = "${artist.name} · Artist",
                target = SearchSuggestionTarget.Artist(artist.id),
            )
    }
    for (album in state.albums) {
        if (suggestions.size >= SEARCH_SUGGESTION_CAP) return suggestions
        suggestions +=
            SearchSuggestionUi(
                id = "album:${album.id}",
                title = album.name,
                label = "${album.name} · Album",
                target = SearchSuggestionTarget.Album(album.id),
            )
    }
    for (track in state.tracks) {
        if (suggestions.size >= SEARCH_SUGGESTION_CAP) return suggestions
        val albumId = track.albumId ?: continue
        suggestions +=
            SearchSuggestionUi(
                id = "track:${track.id}",
                title = track.title,
                label = "${track.title} · Track",
                target = SearchSuggestionTarget.TrackAlbum(albumId),
            )
    }

    return suggestions
}

File: apps/web/src/features/search/searchSuggestions.ts
/**
 * Suggestion derivation for `SearchPage.tsx`'s field — `docs/design/screens/SEARCH.md` §6.2.
 * `packages/ui/src/components/SearchField.tsx` already has a complete, tested ARIA-combobox
 * suggestion mechanism (`suggestions`/`onSuggestionSelect`); nothing in the app has ever fed it
 * real data. This module is the "real data" half — it does not fetch anything of its own.
 *
 * Suggestions are built from the same `books`/`series`/`authors`/`podcasts` (Audiobookshelf) and
 * `artists`/`albums`/`tracks` (Jellyfin) arrays `SearchPage.tsx` already computes for the results
 * list below — no new query, no new debounce, so the list updates on exactly the cadence the
 * results list already does.
 *
 * Ordering and cap: concatenate in the fixed kind order the results list itself renders in —
 * Books, Series, Authors, Podcasts, Artists, Albums, Tracks — taking items in the order each
 * source returns them, until 8 total are collected across all kinds combined, then stop. A kind
 * with no candidates simply contributes nothing; slots are never reserved for it.
 *
 * A track with no `albumId` is excluded entirely (same "nowhere to go" rule the full results list
 * already applies to that row) — the caller is expected to have already filtered `sources.tracks`
 * to items that carry a non-null `albumId` (`SearchPage.tsx` does, since it needs the same filter
 * for the results list's own click targets).
 *
 * The label is `"{title} · {Kind}"` — U+00B7 MIDDLE DOT, the same separator convention
 * `ALBUM_DETAIL.md`'s meta line established, not a hyphen or em dash — so this is a literal pin,
 * not a formatting choice made locally.
 */

export type SearchSuggestionKind =
  'Book' | 'Series' | 'Author' | 'Podcast' | 'Artist' | 'Album' | 'Track';

/** One kind's worth of candidates, already reduced to just what this module needs — the caller
 * (`SearchPage.tsx`) maps its richer result-list item shapes down to this before calling in. */
export interface SuggestionSourceItem {
  id: string;
  title: string;
}

export interface SearchSuggestionSources {
  books: SuggestionSourceItem[];
  series: SuggestionSourceItem[];
  authors: SuggestionSourceItem[];
  podcasts: SuggestionSourceItem[];
  artists: SuggestionSourceItem[];
  albums: SuggestionSourceItem[];
  /** Already filtered to items with a non-null `albumId` by the caller — see the module doc
   * comment. `albumId` is carried through separately because a track's own navigation target is
   * its *album*, not itself (`SearchPage.tsx`'s existing click handler does the same thing). */
  tracks: (SuggestionSourceItem & { albumId: string })[];
}

/**
 * One derived suggestion. `id` is unique across the whole list (kind-prefixed, since a book and
 * an artist could otherwise coincidentally share an id) — a stable React key and the value
 * `SearchField` reports back through `aria-activedescendant`. `label` is the exact rendered
 * string (`"{title} · {Kind}"`); `title`/`kind`/`originalId`/`albumId` carry what `onSuggestionSelect`
 * needs to fulfil the selection rule (§6.2: set the field to the plain title, then navigate) —
 * `SearchField`'s own `suggestions` prop is structurally typed as `{ id, label }`, so passing this
 * richer shape through it and reading the extra fields back off the object `onSuggestionSelect`
 * hands back is safe without widening the primitive's own type.
 */
export interface SearchSuggestionEntry {
  id: string;
  kind: SearchSuggestionKind;
  originalId: string;
  /** Set only for `kind === 'Track'` — the album its navigation target actually points at. */
  albumId?: string;
  title: string;
  label: string;
}

const SUGGESTION_CAP = 8;

const SEPARATOR = '·'; // U+00B7 MIDDLE DOT

function labelFor(title: string, kind: SearchSuggestionKind): string {
  return `${title} ${SEPARATOR} ${kind}`;
}

/** Kind order fixed to match the results list below it — Books, Series, Authors, Podcasts,
 * Artists, Albums, Tracks (§6.2). */
export function deriveSearchSuggestions(sources: SearchSuggestionSources): SearchSuggestionEntry[] {
  const ordered: {
    kind: SearchSuggestionKind;
    items: (SuggestionSourceItem & { albumId?: string })[];
  }[] = [
    { kind: 'Book', items: sources.books },
    { kind: 'Series', items: sources.series },
    { kind: 'Author', items: sources.authors },
    { kind: 'Podcast', items: sources.podcasts },
    { kind: 'Artist', items: sources.artists },
    { kind: 'Album', items: sources.albums },
    { kind: 'Track', items: sources.tracks },
  ];

  const result: SearchSuggestionEntry[] = [];
  for (const { kind, items } of ordered) {
    for (const item of items) {
      if (result.length >= SUGGESTION_CAP) return result;
      result.push({
        id: `${kind.toLowerCase()}-${item.id}`,
        kind,
        originalId: item.id,
        albumId: kind === 'Track' ? item.albumId : undefined,
        title: item.title,
        label: labelFor(item.title, kind),
      });
    }
  }
  return result;
}
